File: app/src/main/java/bb/incognito/repositories/GuestRepository.java
package bb.incognito.repositories;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import java.util.List;

import bb.incognito.AppDatabase;
import bb.incognito.dao.GuestDao;
import bb.incognito.model.Guest;
import bb.incognito.model.GuestWithCocktails;

public class GuestRepository {
    private GuestDao guestDao;
    private LiveData<List<Guest>> allGuests;

    public GuestRepository(Application application) {
        AppDatabase db = AppDatabase.getDatabase(application);
        guestDao = db.guestDao();
        allGuests = guestDao.allGuests();
    }

    public LiveData<Guest> getGuest(Integer id) { return guestDao.getGuest(id); }
    public LiveData<List<Guest>> getAllGuests() {
        return allGuests;
    }

    public void update(Guest guest){
        new updateAsyncTask(guestDao).execute(guest);
    }

    public void insert(Guest guest) {
        new insertAsyncTask(guestDao).execute(guest);
    }

    public void delete(Guest guest) {
        new deleteAsyncTask(guestDao).execute(guest);
    }

    private static class insertAsyncTask extends AsyncTask<Guest, Void, Void> {
        private GuestDao guestDao;

        insertAsyncTask(GuestDao guestDao) {
            this.guestDao = guestDao;
        }

        @Override
        protected Void doInBackground(final Guest... params) {
            guestDao.insertGuest(params[0]);
            return null;
        }
    }

    private static class updateAsyncTask extends AsyncTask<Guest, Void, Void> {
        private GuestDao guestDao;

        updateAsyncTask(GuestDao guestDao) {
            this.guestDao = guestDao;
        }

        @Override
        protected Void doInBackground(final Guest... params) {
            guestDao.update(params[0]);
            return null;
        }
    }

    private static class deleteAsyncTask extends AsyncTask<Guest, Void, Void> {
        private GuestDao guestDao;

        deleteAsyncTask(GuestDao guestDao) {
            this.guestDao = guestDao;
        }

        @Override
        protected Void doInBackground(final Guest... params) {
            guestDao.delete(params[0]);
            return null;
        }
    }
}

File: app/src/main/java/bb/incognito/repositories/GuestRepository.kt
package bb.incognito.repositories

import android.app.Application
import android.arch.lifecycle.LiveData
import android.os.AsyncTask

import bb.incognito.AppDatabase
import bb.incognito.dao.GuestDao
import bb.incognito.model.Guest
import bb.incognito.model.GuestWithCocktails

class GuestRepository(application: Application) {
    private val guestDao: GuestDao
    val allGuests: LiveData<List<Guest>>

    init {
        val db = AppDatabase.getDatabase(application)
        guestDao = db.guestDao()
        allGuests = guestDao.allGuests()
    }

    fun getGuest(id: Int?): LiveData<Guest> {
        return guestDao.getGuest(id!!)
    }

    fun update(guest: Guest) {
        updateAsyncTask(guestDao).execute(guest)
    }

    fun insert(guest: Guest) {
        insertAsyncTask(guestDao).execute(guest)
    }

    fun delete(guest: Guest) {
        deleteAsyncTask(guestDao).execute(guest)
    }

    private class insertAsyncTask internal constructor(private val guestDao: GuestDao) : AsyncTask<Guest, Void, Void>() {

        override fun doInBackground(vararg params: Guest): Void? {
            guestDao.insertGuest(params[0])
            return null
        }
    }

    private class updateAsyncTask internal constructor(private val guestDao: GuestDao) : AsyncTask<Guest, Void, Void>() {

        override fun doInBackground(vararg params: Guest): Void? {
            guestDao.update(params[0])
            return null
        }
    }

    private class deleteAsyncTask internal constructor(private val guestDao: GuestDao) : AsyncTask<Guest, Void, Void>() {

        override fun doInBackground(vararg params: Guest): Void? {
            guestDao.delete(params[0])
            return null
        }
    }
}